org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/adapters/ProgressNotificationAdapter.java

```diff
@@ -0,0 +1,40 @@
+/*******************************************************************************
+ * Copyright (c) 2021 1C-Soft LLC and others.
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+ *******************************************************************************/
+package org.eclipse.lsp4j.adapters;
+
+import java.util.function.Predicate;
+
+import org.eclipse.lsp4j.WorkDoneProgressNotification;
+import org.eclipse.lsp4j.jsonrpc.json.adapters.EitherTypeAdapter;
+import org.eclipse.lsp4j.jsonrpc.json.adapters.EitherTypeAdapter.PropertyChecker;
+import org.eclipse.lsp4j.jsonrpc.messages.Either;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonElement;
+import com.google.gson.TypeAdapter;
+import com.google.gson.TypeAdapterFactory;
+import com.google.gson.reflect.TypeToken;
+
+public class ProgressNotificationAdapter implements TypeAdapterFactory {
+
+	private static final TypeToken<Either<WorkDoneProgressNotification, JsonElement>> ELEMENT_TYPE
+			= new TypeToken<Either<WorkDoneProgressNotification, JsonElement>>() {};
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public <T> TypeAdapter<T> create(Gson gson, TypeToken<T> type) {
+		Gson gson2 = gson.newBuilder().registerTypeHierarchyAdapter(WorkDoneProgressNotification.class,
+				new WorkDoneProgressNotificationAdapter()).create();
+		Predicate<JsonElement> leftChecker = new PropertyChecker("kind", "begin").or(
+				new PropertyChecker("kind", "report")).or(new PropertyChecker("kind", "end"));
+		Predicate<JsonElement> rightChecker = t -> true;
+		return (TypeAdapter<T>) new EitherTypeAdapter<>(gson2, ELEMENT_TYPE, leftChecker, rightChecker);
```

Comment:
Does the read/write code in `JsonElementTypeAdapter` need to be used here? I don't know if it matters though as `SemanticTokensClientCapabilitiesRequests.range` does not use `JsonElementTypeAdapter`, but all cases of a field of type `Object` do, such as `InitializeParams.initializationOptions`

Reply:
I'm not sure about it either. All I can tell is that when I was working on this, I was not aware about the case of `SemanticTokensClientCapabilitiesRequests.range`. Therefore, I did it similar to all cases of a field of type `Object`. Also, it seems that if it was done similar to `SemanticTokensClientCapabilitiesRequests.range`, the issue of `? extends Object` mentioned in the previous comment would probably go away.

Reply:
> Also, it seems that if it was done similar to SemanticTokensClientCapabilitiesRequests.range, the issue of ? extends Object mentioned in the previous comment would probably go away.

I have tried it, and no, it would not. This is an unrelated issue, sorry.

Reply:
In case I was not clear or misunderstood your comment, the current implementation of `ProgressNotificationAdapter` will parse a value such as `"value": { "foo": "bar" }` into a `JsonObject` rather than a `LinkedHashMap`. I thought it was correct because I had not known about the case of `SemanticTokensClientCapabilitiesRequests.range`. If it would be better to parse it in a `LinkedHashMap` (e.g., for consistency with `SemanticTokensClientCapabilitiesRequests.range`), please let me know and I'll fix it.

Reply:
I vote to leave it as you have written it now - it is consistent with other uses. If it is wrong, we should fix all the cases in a new issue.

Reply:
> In case I was not clear or misunderstood your comment, the current implementation of `ProgressNotificationAdapter` will parse a value such as `"value": { "foo": "bar" }` into a `JsonObject` rather than a `LinkedHashMap`. I thought it was correct because I had not known about the case of `SemanticTokensClientCapabilitiesRequests.range`. If it would be better to parse it in a `LinkedHashMap` (e.g., for consistency with `SemanticTokensClientCapabilitiesRequests.range`), please let me know and I'll fix it.

I didn't see this comment before I just marked as resolved. Thanks for looking into this more are identifying this difference. I think it should be made consistent and where possible not have `JsonObject`s being returned from the LSP4J API. I don't know if that is a rule or design intention in the past though.

Reply:
> I think it should be made consistent and where possible not have JsonObjects being returned from the LSP4J API. I don't know if that is a rule or design intention in the past though.

I must say I'm a bit puzzled now. I have found issue #126, where `JsonElementTypeAdapter` was introduced specifically for 'parsing Object into a JsonElement'. It looks like the case of `SemanticTokensClientCapabilitiesRequests.range` is an exception rather than a rule. Therefore, I would rather concur with your previous suggestion:

> I vote to leave it as you have written it now - it is consistent with other uses.

WDYT?

Reply:
> I think it should be made consistent and where possible not have JsonObjects being returned from the LSP4J API. I don't know if that is a rule or design intention in the past though.

Yeah - so that wasn't true. Thank you for looking a little deeper and pointing #126.

Reply:
>> I vote to leave it as you have written it now - it is consistent with other uses.
> WDYT?

Yes.